app/scripts/lib/transaction-metrics.ts

```diff
@@ -0,0 +1,879 @@
+import { isHexString } from 'ethereumjs-util';
+import EthQuery from 'eth-query';
+import { BigNumber } from 'bignumber.js';
+import type { Provider } from '@metamask/network-controller';
+
+import { ORIGIN_METAMASK } from '../../../shared/constants/app';
+import {
+  determineTransactionAssetType,
+  isEIP1559Transaction,
+} from '../../../shared/modules/transaction.utils';
+import { hexWEIToDecGWEI } from '../../../shared/modules/conversion.utils';
+import {
+  TransactionType,
+  TokenStandard,
+  TransactionApprovalAmountType,
+  TransactionMetaMetricsEvent,
+  TransactionMeta,
+} from '../../../shared/constants/transaction';
+import { MetaMetricsEventCategory } from '../../../shared/constants/metametrics';
+import { GasRecommendations } from '../../../shared/constants/gas';
+import { TRANSACTION_ENVELOPE_TYPE_NAMES } from '../../../shared/lib/transactions-controller-utils';
+
+///: BEGIN:ONLY_INCLUDE_IN(blockaid)
+import {
+  BlockaidReason,
+  BlockaidResultType,
+} from '../../../shared/constants/security-provider';
+///: END:ONLY_INCLUDE_IN
+
+export const METRICS_STATUS_FAILED = 'failed on-chain';
+
+type ControllerActions = {
+  createEventFragment: (arg0: any) => any;
```

Comment:
I appreciate it's a painful process, but as we're making the huge gain of moving this logic to TypeScript, can we make these callback types as detailed as possible?

Reply:
All sorted in https://github.com/MetaMask/metamask-extension/pull/21330/commits/feca21f8de3cce0ac055c262176a816b6dc06e5a
except `GasFeeState`, that requires some changes in `buildEventFragmentProperties` regarding to the types, noted as a comment